src/node/txpackagetracker.h

```diff
@@ -58,6 +59,36 @@ class TxPackageTracker {
     /** Return how many entries exist in the orphange */
     size_t OrphanageSize();
 
+    /** Received an announcement from this peer for a tx we already know is an orphan; should be
+     * called for every peer that announces the tx, even if they are not a package relay peer.
+     * The orphan request tracker will decide when to request what from which peer - use
+     * GetOrphanRequests().
+     */
+    void AddOrphanTx(NodeId nodeid, const uint256& wtxid, const CTransactionRef& tx, bool is_preferred, std::chrono::microseconds reqtime);
+
+    /** Number of packages we are working on with this peer. Includes any entries in the orphan
+     * tracker, in-flight orphan parent requests (1 per orphan regardless of how many missing
+     * parents were requested), package info requests, tx data download, and packages in the
+     * validation queue. */
+    size_t Count(NodeId nodeid) const;
+
+    /** Number of packages we are currently working on with this peer (i.e. reserving memory for
+     * storing orphan(s)). Includes in-flight package info requests, tx data download, and packages
+     * in the validation queue. Excludes entries in the orphan tracker that are just candidates. */
+    size_t CountInFlight(NodeId nodeid) const;
+
+    /** Get list of requests that should be sent to resolve orphans. These may be wtxids to send
+     * getdata(ANCPKGINFO) or txids corresponding to parents. Automatically marks the orphans as
+     * having outgoing requests. */
+    std::vector<GenTxid> GetOrphanRequests(NodeId nodeid, std::chrono::microseconds current_time);
+
+    /** Update transactions for which we have made "final" decisions: transactions that have
+     * confirmed in a block, conflicted due to a block, or added to the mempool already.
+     * Should be called on new block: valid=block transactions, invalid=conflicts.
+     * Should be called when tx is added to mempool.
+     * Should not be called when a tx fails validation.
+     * */
+    void FinalizeTransactions(const std::set<uint256>& valid, const std::set<uint256>& invalid);
```

Comment:
This is unclear what level of enforcement is granted to a “final” decision, if the transactions are added to any transaction-relay download filters such as `m_recent_rejects` or `m_recently_announced_invs`. Otherwise, in case of blocks re-orgs dropping out the first package component and a second package component still stuck in filters, this can be source of package propagation failure.

Reply:
Note that the rejection filters are cleared every time the chain tip changes, so I don't think we can get failures this way

Reply:
> Note that the rejection filters are cleared every time the chain tip changes, so I don't think we can get failures this way

Yes, I think we should be careful to clean the whole package state in all filters at every block change to avoid “half” state interfering. `m_recent_rejects_reconsiderable_were_rejected` say “reset this filter when the chain tip changes”, of course it includes _backward_ change like re-orgs.